tatanka/client/mesh/trade.go

```diff
@@ -8,77 +8,215 @@ import (
+			Qty:     match.Qty,
+			BaseID:  m.baseID,
+			QuoteID: m.quoteID,
+			Stamp:   now,
+		}
+		o.matches[matchUs.ID()] = matchUs
+		o.remain -= match.Qty
+	}
+	// TODO: Retry with orders from the book if some were not accepted but
+	// more compatible orders exist. We need a way to mark tried orders
+	// first.
+	m.ordsMtx.Lock()
+	m.ords[oid] = o
+	m.ordsMtx.Unlock()
 }
 
-func (m *market) addMatchProposal(match *tanka.Match) {
-	m.ordsMtx.Lock()
-	defer m.ordsMtx.Unlock()
-	ord, found := m.ords[match.OrderID]
-	if !found {
-		m.log.Debugf("ignoring match proposal for unknown order %s", match.OrderID)
+// TODO: Correct these.
+func (m *market) feeParams() *trade.FeeParameters {
+	return &trade.FeeParameters{
+		MaxFeeExposure:    1,
+		BaseFeesPerMatch:  1,
+		QuoteFeesPerMatch: 1,
 	}
-	// Make sure it's not already known or accepted
-	mid := match.ID()
-	if ord.proposed[mid] != nil {
-		// Already known
-		return
+}
+
+func (m *market) addOrder(ord *tanka.Order) {
+	// TODO: Validate order.
+	m.book.Add(ord)
+	check := func(o *order) {
+		o.matchesMtx.Lock()
+		defer o.matchesMtx.Unlock()
+		if o.remain == 0 {
+			return
+		}
+		if ord.Sell == o.Order.Sell {
+			return
+		}
+		if o.Sell {
+			if ord.Rate < o.Order.Rate {
+				return
+			}
+		} else if ord.Rate > o.Order.Rate {
+			return
+		}
+		if compat, _ := trade.OrderIsMatchable(o.remain, ord, m.feeParams()); !compat {
+			return
+		}
+		maxQty := ord.Qty
+		if ord.Qty > o.remain {
+			maxQty = o.remain
+		}
+		lots := maxQty / ord.LotSize
+		qty := lots * ord.LotSize
+		o.remain -= qty
+		now := time.Now()
+		matchThem := &tanka.Match{
+			From:    m.peerID,
+			OrderID: o.Order.ID(),
+			Qty:     qty,
+			BaseID:  m.baseID,
+			QuoteID: m.quoteID,
+			Stamp:   now,
+		}
+		ok, err := m.negotiate(ord.From, matchThem)
+		if err != nil {
+			m.log.Errorf("unable to negotiate match with peer %s: %v", ord.From, err)
+			return
+		}
+		if !ok {
+			return
+		}
+		// They agree.
+		matchUs := &tanka.Match{
+			From:    ord.From,
+			OrderID: ord.ID(),
+			Qty:     qty,
+			BaseID:  m.baseID,
+			QuoteID: m.quoteID,
+			Stamp:   now,
+		}
+		o.matches[matchUs.ID()] = matchUs
 	}
-	if ord.accepted[mid] != nil {
-		// Already accepted
-		return
+	m.ordsMtx.RLock()
+	defer m.ordsMtx.RUnlock()
+	for _, ord := range m.ords {
+		check(ord)
 	}
-	ord.proposed[mid] = match
 }
 
-func (m *market) addMatchAcceptance(match *tanka.Match) {
-	m.ordsMtx.Lock()
-	defer m.ordsMtx.Unlock()
+func (m *market) negotiate(to tanka.PeerID, match *tanka.Match) (bool, error) {
+	b, err := json.Marshal(*match)
+	if err != nil {
+		return false, err
+	}
+	msg := &msgjson.Message{
+		Route:   mj.RouteNegotiate,
+		Payload: b,
+	}
+	// Is this necessary or will request try the connection first?
+	if err := m.conn.ConnectPeer(to); err != nil {
+		return false, err
+	}
+	var ok bool
+	return ok, m.conn.RequestPeer(to, msg, &ok)
+}
+
+func (m *market) handleNegotiate(match *tanka.Match, respond func(any, *msgjson.Error)) {
```

Comment:
I don't think we should pass `respond` into here. The caller should pass whatever is returned from this function into `respond`.

Reply:
Ok, this just returning a bool now.